sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -543,15 +602,19 @@ def process_bundle_split(self,
                            instruction_id  # type: str
                           ):
     # type: (...) -> beam_fn_api_pb2.InstructionResponse
-    processor = self.bundle_processor_cache.lookup(request.instruction_id)
-    if processor:
-      return beam_fn_api_pb2.InstructionResponse(
-          instruction_id=instruction_id,
-          process_bundle_split=processor.try_split(request))
-    else:
+    try:
+      processor = self.bundle_processor_cache.lookup(request.instruction_id)
+    except RuntimeError:
       return beam_fn_api_pb2.InstructionResponse(
-          instruction_id=instruction_id,
-          error='Instruction not running: %s' % instruction_id)
+          instruction_id=instruction_id, error=traceback.format_exc())
+    # Return an empty response if we aren't running. This can happen
+    # if the ProcessBundleRequest has not started or already finished.
+    process_bundle_split = (
+        processor.try_split(request)
+        if processor else beam_fn_api_pb2.ProcessBundleSplitResponse())
```

Comment:
So, the correct response if something hasn't previously failed, and isn't running is an empty response for both Splits and Progress requests?

That's easy enough to fix. Go current fails, but it's very easy to swap that up.

Reply:
Yes, but you should strive to return an error for unknown/failed instructions.

This PR also keeps a fixed number of completed/failed instructions in a LRU like object so that even after the bundle completes we can still return a useful answer before we start sending the `I don't know` answer.

Reply:
So to be clear, the semantics are required to be:

Receive PB request for INST. -> Empty Splits & and Progress
INST starts having results -> Return results.
INST finishes -> EmptySplits and progress until "expired".
And the "I don't know" is an error for  every other circumstance?

In the case where the bundle finishes, but requires Finalization, it INST should remain valid and return the it's current set of responses for splits and progress as usual, rather than the empties.

Reply:
Filed https://issues.apache.org/jira/browse/BEAM-10976 for implementing Go SDK bundle finalization along with a reminder about this issue regarding the active status.

Reply:
You'll want to differentiate the failed case and once the Go SDK handles bundle finalization consider that as well (i.e. keep INST live until the finalize request comes or a user specified timeout happens)

Reply:
The Go SDK already differentiates the failed case and re-forwards the cached error.